### ISSUE TITLE
onNotificationReceived stop working after a while

### ISSUE THREAD

**Issue by smartls** [leandrosimoes/react-native-android-notification-listener#4]
I put the library in my app, at first it worked properly. But after a moment(about 20 mins), onNotificationReceived no longer be called. I suspect the service was recycled by OS, after some google, I found same issues [here](https://gist.github.com/xinghui/b2ddd8cffe55c4b62f5d8846d5545bf9).
As the writer said, NLS became a zombie service, which caused onNotificationReceived losing its function. I also test it with adb command:`adb shell dumpsys notification`, found the app disappear from the "Live Notification Listeners" section.
I'm a newbie in native android, do you have any suggestions?

**Comment by leandrosimoes** [comment]
Hi @smartls, thanks for open this issue. I'll take a look, do some tests and let you know asap.

**Comment by smartls** [comment]
@leandrosimoes thanks for your working. after some hours tests, I add a new method to "wake up" NotificationLisenerService, invoke it when app loads. May be it's not a best practise, but works fine for me.
`
    @ReactMethod
    public void ensureListenerRunning(){
        ComponentName collectorComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
        ActivityManager manager = (ActivityManager) reactContext.getSystemService(Context.ACTIVITY_SERVICE);
        boolean collectorRunning = false;
        List<ActivityManager.RunningServiceInfo> runningServices = manager.getRunningServices(Integer.MAX_VALUE);
        if (runningServices == null ) {
            return;
        }
        for (ActivityManager.RunningServiceInfo service : runningServices) {
            if (service.service.equals(collectorComponent)) {
                if (service.pid == Process.myPid()) {
                    collectorRunning = true;
                }
            }
        }
        if (collectorRunning) {
            return;
        }
        toggleNotificationListenerService();
    }

    private void toggleNotificationListenerService(){
        ComponentName thisComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
        PackageManager pm = reactContext.getPackageManager();
        pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
        pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
    }

`

**Comment by c7nasr** [comment]
> @leandrosimoes thanks for your working. after some hours tests, I add a new method to "wake up" NotificationLisenerService, invoke it when app loads. May be it's not a best practise, but works fine for me.
> 
> ```java
> @ReactMethod
> public void ensureListenerRunning(){
>     ComponentName collectorComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
>     ActivityManager manager = (ActivityManager) reactContext.getSystemService(Context.ACTIVITY_SERVICE);
>     boolean collectorRunning = false;
>     List<ActivityManager.RunningServiceInfo> runningServices = manager.getRunningServices(Integer.MAX_VALUE);
>     if (runningServices == null ) {
>         return;
>     }
>     for (ActivityManager.RunningServiceInfo service : runningServices) {
>         if (service.service.equals(collectorComponent)) {
>             if (service.pid == Process.myPid()) {
>                 collectorRunning = true;
>             }
>         }
>     }
>     if (collectorRunning) {
>         return;
>     }
>     toggleNotificationListenerService();
> }
> 
> private void toggleNotificationListenerService(){
>     ComponentName thisComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
>     PackageManager pm = reactContext.getPackageManager();
>     pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
>     pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
> }
> ```

hey @smartls . 
I'm facing the same issue with Huawei/Honor Phone. does this code edit tested on Huawei Devices? and could you provide me with file path and entire file content ? thanks in advance

**Comment by smartls** [comment]
@c7nasr 
Yes, I tested it on Huawei P20 and Hornor30, you can call ensureListenerRunning interval to keep service activated. This is my whole file.
`public class RNAndroidNotificationListenerModule extends ReactContextBaseJavaModule {
    private static ReactApplicationContext reactContext;
    private static final String TAG = "RNAndroidNotificationListener";

    public RNAndroidNotificationListenerModule(ReactApplicationContext context) {
        super(context);

        reactContext = context;
    }

    @Override
    public String getName() {
        return TAG;
    }

    public static void sendEvent(String event, WritableMap params) {
        if (reactContext == null) return;

        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(event, params);
    }

    @ReactMethod
    public void getPermissionStatus(Promise promise) {
        if (reactContext == null) return;

        String packageName = reactContext.getPackageName();
        Set<String> enabledPackages = NotificationManagerCompat.getEnabledListenerPackages(reactContext);
        if (enabledPackages.contains(packageName)) {
            promise.resolve("authorized");
        } else {
            promise.resolve("denied");
        }
    }

    @ReactMethod
    public void requestPermission() {
        if (reactContext == null) return;

        final Intent i = new Intent();
        i.setAction(Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS);
        i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        i.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
        i.addFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS);
        reactContext.startActivity(i);
    }

    @ReactMethod
    public void ensureListenerRunning(){
        ComponentName collectorComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
        ActivityManager manager = (ActivityManager) reactContext.getSystemService(Context.ACTIVITY_SERVICE);
        boolean collectorRunning = false;
        List<ActivityManager.RunningServiceInfo> runningServices = manager.getRunningServices(Integer.MAX_VALUE);
        if (runningServices == null ) {
            return;
        }
        for (ActivityManager.RunningServiceInfo service : runningServices) {
            if (service.service.equals(collectorComponent)) {
                if (service.pid == Process.myPid()) {
                    collectorRunning = true;
                }
            }
        }
        if (collectorRunning) {
            return;
        }
        toggleNotificationListenerService();
    }

    private void toggleNotificationListenerService(){
        ComponentName thisComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
        PackageManager pm = reactContext.getPackageManager();
        pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
        pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
    }
}`

**Comment by c7nasr** [comment]
Thanks for your response.
How did you call ensure Listener running interval. I tried background fetch but after 4 to 7 hours it doesn't keep working

**Comment by smartls** [comment]
@c7nasr 
4-7 hours? really a long time, I didn't try so long running of the app, in my opinion, app always be disposed in several hours by OS.
I just call the method every 20 mins.

**Comment by leandrosimoes** [comment]
@c7nasr First of all, thanks a lot for your answers here! Could you please create a PR with this code of yours together with proper documentation on the README showing how to use that? If somehow this will not be possible, could you please provide for me just the documentation about how to use so I could create that in and publish?

**Comment by c7nasr** [comment]
> @c7nasr First of all, thanks a lot for your answers here! Could you please create a PR with this code of yours together with proper documentation on the README showing how to use that? If somehow this will not be possible, could you please provide for me just the documentation about how to use so I could create that in and publish?

Let me clarify the situation,

I have an app that depends on Incoming notifications 100%, so I used this package to listen to incoming notifications. but after closing the app **about 10 to 15 minutes** the listener is gone .. (Honor 10Lite)

So to resolve that issue, I used background fetch library to check if the notification listener existed every 15 minutes. But surprise! Android kills background fetch :( after long time (10-12) hours. So I lose the whole functionality of my app!

That's disappointing me. But when I tried to implement the notification listener natively (Without Packages). surprise, Android didn't kill the listener even after a long time. _**I don't know why. but its work!**_

**Comment by leandrosimoes** [comment]
Understood @c7nasr , thanks again for the answer, I'll try to take a look on that. 

@smartls could you please provide more information about how you use this code at the RN side to keep the listener running? And thanks a lot for your answers too!

**Comment by c7nasr** [comment]
I'll post any updates **if really native notification listener works for long periods** like days.
I have pushbullet app it ** the same idea**. I barely open it. But their notification listener is really awesome it running 24/7!

So I think native listener gonna make it. but I'll keep you updated .

**Comment by leandrosimoes** [comment]
Awesome @c7nasr ... if that really works, could you please provide the source code so we can see how you implemented this notification listener? It would be awesome if I could make it work here too. Thanks a lot.

**Comment by c7nasr** [comment]
![image](https://user-images.githubusercontent.com/42423651/93202926-d40a1c00-f753-11ea-88d3-194132d2f677.png)

Good News. It still **LIVE**; the system doesn't kill the service. Honestly, your package service isn't killed too, but it doesn't trigger the function that should be triggered.
In my opinion, Notification Listener could be only **NATIVE** implementation not through the react-native bridge

**Comment by lucaswhitman** [comment]
@c7nasr I'd love to see some code on how you did this. We're struggling with the same thing.

**Comment by c7nasr** [comment]
I just declared the notification listener natively **without packages**. And i made a service that starts when a notification received. And i shutdown the service from JS handless function after complete the task. 

If anyone want code just tell me share gist with you

**Comment by lucaswhitman** [comment]
I'd love to see a gist. It would certainly be helpful, we're getting ready to prioritize a fix here.

**Comment by leandrosimoes** [comment]
That would be great @c7nasr !

**Comment by jehartzog** [comment]
@c7nasr We like to move quickly and tried to implement this based on your description.

[here] is our fork, we switched from sending events over the RN bridge, and directly sending them to a standalone `HeadlessJsTaskService`. The RN module portions are now only there for permissions.

Is this what you found worked for you? The only comment I couldn't quite follow from your description was:

> And i shutdown the service from JS handless function after complete the task

Since AFAIK the headlessJs service will automatically shutdown when the RN `async` task returns, or when the timeout is reached.

Please let me know if you have any feedback. We're going to be testing this over a few days (since we're more interested in longer-term stable monitoring), and if so will organize our code & docs a little more to get a PR ready. This solution does require some minor changes to the RN side that aren't yet shown in this fork.

**Comment by aftabaminzoobiapps** [comment]
Hi. Did you solve this issue.

**Comment by aftabaminzoobiapps** [comment]
> @leandrosimoes thanks for your working. after some hours tests, I add a new method to "wake up" NotificationLisenerService, invoke it when app loads. May be it's not a best practise, but works fine for me.
> 
> ```java
> @ReactMethod
> public void ensureListenerRunning(){
>     ComponentName collectorComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
>     ActivityManager manager = (ActivityManager) reactContext.getSystemService(Context.ACTIVITY_SERVICE);
>     boolean collectorRunning = false;
>     List<ActivityManager.RunningServiceInfo> runningServices = manager.getRunningServices(Integer.MAX_VALUE);
>     if (runningServices == null ) {
>         return;
>     }
>     for (ActivityManager.RunningServiceInfo service : runningServices) {
>         if (service.service.equals(collectorComponent)) {
>             if (service.pid == Process.myPid()) {
>                 collectorRunning = true;
>             }
>         }
>     }
>     if (collectorRunning) {
>         return;
>     }
>     toggleNotificationListenerService();
> }
> 
> private void toggleNotificationListenerService(){
>     ComponentName thisComponent = new ComponentName(reactContext, RNAndroidNotificationListener.class);
>     PackageManager pm = reactContext.getPackageManager();
>     pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
>     pm.setComponentEnabledSetting(thisComponent, PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
> }
> ```

In which file we have to add these lines.

**Comment by leandrosimoes** [comment]
Hi @jehartzog ... I took a look at your fork and your implementation and I'm a little bit confused about what you said that `The RN module portions are now only there for permissions`. Just to clarify, the method `onNotificationReceived` in the JS side is still being called when a notification is received? I'm asking because If it is the case, and the service is being 24/7 active after all, I think that would be great to get these changes merged in this repo.

**Comment by leandrosimoes** [comment]
Hi guys, I made some changes based on the @jehartzog implementations of the Headless JS service. I also add a bootup receiver to start the service in case the user restarts the device too. I just need to do more tests and check if everything will be fine for a long period of running and I will be really greatefull if someone could please help me with that, just use the branch `fix-service-becoming-zombie` that is the branch that I open the PR and let me know if everything is alright.

I'll wait some days before merging this PR to get more feedback about the changes from you guys.

I really hope that the problem solves with these changes. 

Thanks a lot for the messages.

**Comment by jehartzog** [comment]
@leandrosimoes Thanks for working to bring the changes into the repo. I'm generally willing to put together an actual PR, but I've learned to ask if the author is at all interested in merging before putting in that effort, lol.

About your question:
> I'm a little bit confused about what you said that The RN module portions are now only there for permissions

What I meant is that [this module](https://github.com/leandrosimoes/react-native-android-notification-listener/blob/fix-service-becoming-zombie/android/src/main/java/com/lesimoes/androidnotificationlistener/RNAndroidNotificationListenerModule.java#L15-L56) only uses the RN bridge to handle permissions now. For actual long running notification event data, we use the HeadlessJS process. I've reviewed your branch briefly and it matches up with my recommended changes.

I will add that we've been using this in production for months now, and it is extremely stable. I believe that the bootup receiver may be unnecessary based on our testing, but likely wouldn't hurt at all.

Here are my only recommendations to your branch:
1. The JS headless receiver functions needs to be `async`. This will ensure the task is not `paused` while the promise is still being executed.
```
const headlessNotificationListener = notification => {
```ts
change to
```ts
const headlessNotificationListener = async (notification) => {
```
2. You should really link directly to [https://reactnative.dev/docs/headless-js-android](https://reactnative.dev/docs/headless-js-android) in the readme, as there are more nuances to using headlessJs (such as you can't directly touch UI inside of it). Don't bother trying to copy the important stuff, just warn users they need to review it. I suspect a lot of RN devs will not have directly worked with headlessJs on Android before. You may want to add a comment in the readme example of the headlessJS code that this is JS running in a standalone 'task' and can't directly touch UI, to help warn.

**Comment by leandrosimoes** [comment]
Thanks, @jehartzog ! I did the changes that you said in the documentation. Could you please take a look and see if it is good enough?

**Comment by leandrosimoes** [comment]
Released version v2.0.0 that should fix this problem. WARNING: There are some breaking changes in this version, see the documentation in README!!!
